### ISSUE TITLE
[Feature] Add support for Custom Tools Executors

### ISSUE THREAD

**Issue by RobertDiebels** [nx-dotnet/nx-dotnet#455]
**Problem description**
At one of my clients we use Nx to build a mono-repo. We use Nx-dotnet to build some of our applications and we hit a roadblock. We wanted to add QA measures to our builds using SonarScanner. 

Unfortunately, SonarScanner for .NET does not play nice with Nx as it requires that each project's `dotnet build` command is wrapped with a `dotnet sonarscanner /begin` and a `dotnet sonarscanner /end`.
Furthermore, nx-dotnet does not make it easy to add your own Custom Executors for `dotnet tools`. I had to do some dirty hacking and reuse some of your code that is not exposed in your API to get it to work.  

In short the problem is two-fold:
- Nx-dotnet does not facilitate its executors be wrapped by pre- and post-executors.
- Nx-donet does not facilitate Custom .NET Tools Executors be build using its API.

**Possible Solution**

_.NET Tools Exectuors_

Primarily, I would like to request that the `nx-dotnet` API be extended to allow users to write their own executors for extensions to the .NET build-system/cli. Primarily those extensions added using `dotnet tools`. 

This would make your API more flexible and for instance allow you to treat the `dotnet format` (which is a dotnet tool) as a separate component and remove its associated method from the `DotNetClient` interface.

_Wrapping dotnet commands_

I would like to request that  nx-dotnet facilitates wrapping `dotnet X` commands with custom .NET tools executors. This would allow the use-case as described above to be implemented with greater ease.

**Alternatives**

Alongside this Feature request I will be filing an issue with SonarSource to make their SonarScanner for .NET compatible with mono-repo build-systems. Since their scanner cannot co-operate with build-systems that build multiple projects under one command. [ [Feature-request](https://community.sonarsource.com/t/support-monorepo-build-systems-for-net-sonarscanner-azure-pipelines-tasks/64579) ]

**Additional context**
The mono-repo project is built on Azure DevOps and their are no tasks for Nx or SonarScanner that allow me to do what I need to do.

**Comment by AgentEnder** [comment]
First, some things that we **_are_** planning to do in nx-dotnet:
- run-tool executor
- msbuild task executor

The run-tool executor would allow you to configure Nx's targetDependencies to get you to where you need to be, but I don't think it would be ideal for a few reasons. 
- You'd miss out on nx-cloud's DTE feature. The running of the sonarscan tool produces a side effect that would only be on the agent machine that it ran on, so the build couldn't be distributed. While not everyone uses DTE (currently, most people haven't adopted it), it would be bad design to introduce solutions that wouldn't work well with it
- You'd need to configure targetDependencies to set this up, and most people don't delve way too deep into that either.

My proposal would be to add prebuild and postbuild msbuild targets in your .csproj files that use exec to run these commands. As an example, there is already a prebuild task added if you generate a project with nx-dotnet.

If you want this for every new project generated, I'd recommend a workspace-generator (nx), local (nx) plugin generator, or custom .NET template that you can use to ensure all projects are created in this manner. If using an Nx generator, you can just wrap the nx-dotnet generators and perform the updates at the end.

**Comment by RobertDiebels** [comment]
@AgentEnder 

>First, some things that we are planning to do in nx-dotnet:

Ah that's good news! Thanks for the heads up :D !

> The run-tool executor would allow you to configure Nx's targetDependencies to get you to where you need to be, but I don't think it would be ideal for a few reasons.

I considered using targetDependencies but, don't Nx executors execute for all projects before the next target is processed? So even if I would introduce a target-dependency this would not solve the "wrapping" of the build command that needs to occur.

> My proposal would be to add prebuild and postbuild msbuild targets in your .csproj files that use exec to run these commands. 

Wouldn't adding prebuild and postbuild tasks just shift the problem onto MSBuild instead of Nx? 

I assume this is the case because, I'm using information from Nx to auto-generate the project-keys that SonarQube/SonarCloud needs to store an analysis of a particular project. This way I do not have to think about adding a project to SonarQube/SonarCloud everytime we add a project to Nx. The Nx executor handles that for us.

If I used the MSBuild tasks I would need that information. I could possibly generate the project-key before running the Exec task and use it as input. Or, create a custom task for MSBuild and use that to generate the keys. Either way I'd be doing a lot of work to make this happen. In Nx or MSBuild.

I'm hoping that SonarSource will pick up on my feature request and allow the scanners to register multi-project builds. Which would no longer require anything from Nx. Though I'm doubtful that will occur any time soon. Either way, thanks for the quick response!

**Comment by AgentEnder** [comment]
> Wouldn't adding prebuild and postbuild tasks just shift the problem onto MSBuild instead of Nx?

Yeah, it kind of would. This is ideal though for **_most_** pre/post build tasks though, since you generally want them to function if calling the build from visual studio / rider / some other IDE that isn't configured to use the Nx commands. In general, I'm trying to target interopt w/ peoples existing dev tools for the plugin, so pre/post build tasks that only execute through Nx make less sense to me.

For sonarscan in particular, this may not be the case. You may only want them on CI, and that's fine. If that is the case, I'd recommend adding a target to your project like this:
```jsonc
{
  "targets": {
    "build": {  }, // <-- Existing nx-dotnet target
    "sonar-build": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "dotnet sonarscan ....",
          "nx build ....",
          "dotnet sonarscan ......",
        ]
      }
    }
}
```

> If I used the MSBuild tasks I would need that information. I could possibly generate the project-key before running the Exec task and use it as input. Or, create a custom task for MSBuild and use that to generate the keys. Either way I'd be doing a lot of work to make this happen. In Nx or MSBuild.

I'd recommend just using a static key for the project, for example the project name. This is likely what you are using anyways 😄.
Wrapping the nx-dotnet generator, you'd specify the key in the msbuild exec task but since that is handled when you add the project through the generator then you don't have to worry about manually setting it up.

Looking at the code example for adding the dep-check rebuild task, we add a "project-key" that matches the nx project name at the end of the task's command.
https://github.com/nx-dotnet/nx-dotnet/blob/64a3e7fd9051a581d56bcd742bacd90e9a4cceb6/packages/core/src/generators/utils/generate-project.ts#L214-L233

**Comment by RobertDiebels** [comment]
> For sonarscan in particular, this may not be the case. You may only want them on CI, and that's fine. If that is the case, I'd recommend adding a target to your project like this:

I've currently implemented a variation of this. Instead of calling the commands directly I've written 3 custom Nx Executors. One for the `dotnet sonarscanner /begin` command, one for the `dotnet sonarscanner /end` command and one that executes the dotNetBuildExecutor with the previous two executors as pre- and post tasks.

This allowed me to provide options and overwrite the project-key, or any other setting for the sonar-scanner if needed.

However, writing the executors was a bit troublesome. That's what prompted me to write this FR. I think it would be beneficial if it were made possible to extend nx-dotnet's API and enable custom Nx .NET Tools executors. This would also allow you to outsource tools-support to other developers.

If the run-tool executor would be for exactly that purpose then I think the major part of this FR would be covered.
I'd like to suggest that you also look at nx-dotnet's API and see how you can improve it or, alternatively disallow it. As it is currently possible to use the `DotNetClient` but it's API is not very intuitive.

Again, thank you for the quick response and exchange of thoughts. Much appreciated!
